source/time_tables/route_schedules.cpp

```diff
@@ -177,37 +173,124 @@ std::vector<Edge> create_edges(std::vector<std::vector<datetime_stop_time>>& v)
-        return order;
-    } catch (boost::not_a_dag&) {}
+    for (const auto& edge: edges) { add_edge(edge.source, edge.target, g); }
+    if (is_dag(g)) {
+        LOG4CPLUS_DEBUG(logger, "total topological sort is working, great!");
+        return std::move(is_dag.order);
+    }
 
     // there is a cycle, thus we run the complete algorithm.
     g = Graph(nb_vertices);// remove all edges
-    for (const auto& edge: edges) {
-        const auto e_desc = add_edge(edge.source, edge.target, g).first;
-        try {
-            order.clear();
-            boost::topological_sort(g, std::back_inserter(order));
-        } catch (boost::not_a_dag&) {
-            remove_edge(e_desc, g);
+    int nb_removed_edges = 0;
+    LOG4CPLUS_DEBUG(logger, "trying total ranked pair with nb_edges = " << edges.size());
+
+    // Straintforward implementation of ranked pair is:
+    //
+    //for (const auto& edge: edges) {
+    //    const auto e_desc = add_edge(edge.source, edge.target, g).first;
+    //    if (! is_dag(g)) { ++nb_removed_edges; remove_edge(e_desc, g); }
+    //}
+    //
+    // Let n be the number of vertices, and e be the number of
+    // edges. O(e) = O(n²).  The straintforward implementation is
+    //
+    //   O(e * (n + e)) = O(e²) = O(n⁴)
+    //
+    // In practice, this algorithm is too costy for big route
+    // schedules.  Let k be the number of edges that will be removed.
+    // O(k) = O(n²).  In practice, it seems that Omega(k) = Omega(n).
+    // If we do binary search to find each edges that we need to
+    // remove, the complexity is
+    //
+    //   O(k * log(e) * (n + e)) = O(k * n² * log(n²)) = O(k * n² * log(n))
+    //
+    // Thus, in our case, this algorithm seems to run in
+    //
+    //   Omega(k * n² * log(n)) = Omega(n³ * log(n))
+    //
+    // which is much better than the naive implementation.  In
+    // practice, our problematic case run in 33s using the naive
+    // algorithm, and 1s using the binary search.
+    //
+    size_t done_until = 0;
+    // the edges in the graph that we may remove during the current binary search
+    std::vector<Graph::edge_descriptor> e_descrs;
+    e_descrs.reserve(edges.size());
+    while (done_until < edges.size()) { // while not all edges are done
+        // binary searching the next edge that create a cycle,
+        // inspired by
+        // https://en.wikipedia.org/wiki/Binary_search_algorithm#Deferred_detection_of_equality
+        e_descrs.clear();
+        size_t imin = done_until;
+        size_t imax = edges.size();
+        while (imin < imax) {
+            const size_t imid = (imin + imax) / 2;
+
+            // modifying the graph to be in the wanted state
+            const size_t nb_elt_to_add = imid - done_until + 1;
+            if (e_descrs.size() < nb_elt_to_add) {
+                // add corresponding edges
+                while (nb_elt_to_add != e_descrs.size()) {
+                    const size_t idx = done_until + e_descrs.size();
```

Comment:
Really not, we have clarified that together and make some modifications, but not here, thus github let the comment.